Document/0x04g-Testing-Cryptography.md

```diff
@@ -69,77 +89,93 @@ See also the following best practice documents for recommendations:
-    - if key wrapping scheme is used, ensure that the master secret is initialized for each user, or container is re-encrypted with new key;
-    - check how password change is handled and specifically, if you can use master secret or previous password to decrypt the container.
+If the app relies on an additional encrypted containers stored in app data, check how the encryption key is used. If a key-wrapping scheme is used, ensure that the master secret is initialized for each user or the container is re-encrypted with new key. If you can use the master secret or previous password to decrypt the container, check how password changes are handled.
 
-Whenever symmetric cryptography is used in mobile apps, the associated secret keys must be stored in secure device storage. For mote information on the platform-specific APIs, refer to the "Testing Data Storage" chapters.
+Secret keys must be stored in secure device storage whenever symmertic cryptography is used in mobile apps. For more information on the platform-specific APIs, see the [**Testing Data Storage on Android**](https://github.com/OWASP/owasp-mstg/blob/master/Document/0x05d-Testing-Data-Storage.md) and [**Testing Data Storage on iOS**](https://github.com/OWASP/owasp-mstg/blob/master/Document/0x06d-Testing-Data-Storage.md) chapters.
 
 ##### Weak Key Generation Functions
 
-Cryptographic algorithms, such as symmetric encryption or MACs, expect a secret input of a given size, e.g. 128 or 256 bit. A native implementation might use the user-supplied password directly as an input key. There are a couple of problems with this approach:
+Cryptographic algorithms (such as symmetric encryption or MACs) expect a secret input of a given size (for example, 128 or 256 bit). A native implementation might use the user-supplied password directly as an input key. Using a user-supplied password as an input key has the following problems:
 
-- If the password is smaller than the key, then not the full key-space is used (the rest is padded, sometimes even with spaces)
-- A user-supplied password will realistically consist mostly of displayable and pronounceable characters. So instead of the full entropy, i.e. 2<sup>8</sup> when using ASCII, only a small subset is used (approx. 2<sup>6</sup>).
-- If two users select the same password an attacker can match the encrypted files. This opens up the possibility of rainbow table attacks.
+- If the password is smaller than the key, the full key space isn't used. The remaining space is padded (spaces are sometimes used for padding).
+- A user-supplied password will realistically consist mostly of displayable and pronounceable characters. Therefore, only a subset of all possible ASCII characters are used and entropy is decresed (from 2<sup>8</sup> to approximately 2<sup>6</sup>).
+- If two users select the same password, an attacker can match the encrypted files. This makes rainbow table attacks a possibility.
```

Comment:
> If two users select the same password, an attacker can match the encrypted files.

This doesn't always hold up. For example, this is not possible if we used a random initialisation vector for each file.

> This makes rainbow table attacks a possibility.

While this may be true in theory, but it wouldn't be practical for encrypted data, since you wouldn't know the data in advance.

Reply:
@b-mueller I deleted this bullet point in the list due to this feedback. Let me know if it should instead stick around but with some corrections